drivers: i2c: stm32_Slave: Fix addr flag handling

pavlohamov opened the pull request:
In the main Addr handler code the F1 workaround was used.
Add compile time swith depending on SOC family.
So workaround is not afffecting F2/F4 families.

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

Commit:
drivers: i2c: stm32_Slave: Fix addr flag handling

In the main Addr handler code the F1 workaround was used.
Add compile time swith depending on SOC family.
So workaround is not afffecting F2/F4 families.

Signed-off-by: Pavlo Hamov <pavlo_hamov@jabil.com>

pavlohamov commented:
It's just for test. Please Do no merge.

avisconti commented on drivers/i2c/i2c_ll_stm32_v1.c:
```diff
@@ -210,6 +210,7 @@ static inline void handle_addr(struct device *dev)
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32F1X
 	if (data->current.is_write || data->current.len > 2) {
```
It seems that in F1 case you need to do a different process than in other cases. I would like to ask you
to describe in the commit message what is happening here from I2C protocol standpoint. Are you sure that
this different behaviour is only for F1? Maybe having a clear vision from I2C protocol
standpoint would help us to understand what is going on on other SoC series.

pavlohamov replied:
Yep. In the F1 family there are some silicon limitations due to errata. As you may see there is a couple of corner cases for data count below 3. From the perspective of i2c proto there nothing to do. I's just a workaround recommended by the errata.

avisconti replied:
> Yep. In the F1 family there are some silicon limitations due to errata. As you may see there is a couple of corner cases for data count below 3. From the perspective of i2c proto there nothing to do. I's just a workaround recommended by the errata.

Can you mention the errata link in the commit message? Or maybe in a comment in the code.

pavlohamov replied:
Done. Link is too long https://www.st.com/content/ccc/resource/technical/document/errata_sheet/f5/50/c9/46/56/db/4a/f6/CD00197763.pdf/files/CD00197763.pdf/jcr:content/translations/en.CD00197763.pdf
So i just add a reference name of ducument.

ydamigos replied:
STM32F103 reference manual [RM0008](https://www.st.com/content/ccc/resource/technical/document/reference_manual/59/b9/ba/7f/11/af/43/d5/CD00171190.pdf/files/CD00171190.pdf/jcr:content/translations/en.CD00171190.pdf) and STM32F401 reference manual [RM0368](https://www.st.com/content/ccc/resource/technical/document/reference_manual/5d/b1/ef/b2/a1/66/40/80/DM00096844.pdf/files/DM00096844.pdf/jcr:content/translations/en.DM00096844.pdf) describe differently the master reception for N < 3.

F1: 
```
The cases where a single byte or two bytes are to be received should be handled differently, as described below:
Case of a single byte to be received:
- In the ADDR event, clear the ACK bit.
- Clear ADDR
- Program the STOP/START bit
- Read the data after the RxNE flag is set.

Case of two bytes to be received:
- Set POS and ACK
- Wait for the ADDR flag to be set
- Clear ADDR
- Clear ACK
- Wait for BTF to be set
- Program STOP
- Read DR twice
```

F4:
```
For 2-byte reception:
- Wait until ADDR = 1 (SCL stretched low until the ADDR flag is cleared)
- Set ACK low, set POS high
- Clear ADDR flag
- Wait until BTF = 1 (Data 1 in DR, Data2 in shift register, SCL stretched low until a data 1 is read)
- Set STOP high
- Read data 1 and 2
```
@pavlohamov Does the "else" part work for F1 family if we just handle the 0 length case:
```
if (data->current.is_write) {
	LL_I2C_ClearFlag_ADDR(i2c);
	return;
}
if (data->current.len == 0U) {
	LL_I2C_ClearFlag_ADDR(i2c);
} else if (data->current.len == 1U) {
	/* Single byte reception: enable NACK and clear POS */
	LL_I2C_AcknowledgeNextData(i2c, LL_I2C_NACK);
} else if (data->current.len == 2U) {
	/* 2-byte reception: enable NACK and set POS */
	LL_I2C_AcknowledgeNextData(i2c, LL_I2C_NACK);
	LL_I2C_EnableBitPOS(i2c);
}
LL_I2C_ClearFlag_ADDR(i2c);
```

avisconti replied:
> Done. Link is too long https://www.st.com/content/ccc/resource/technical/document/errata_sheet/f5/50/c9/46/56/db/4a/f6/CD00197763.pdf/files/CD00197763.pdf/jcr:content/translations/en.CD00197763.pdf
> So i just add a reference name of document.

Yes, name is fine. But pls add also the section in the errata for the workaround.

avisconti replied:
> @pavlohamov Does the "else" part work for F1 family if we just handle the 0 length case:

I see your point.
0 and 1 should be handled in the same way. I'm not able to find the case in the errata though.

pavlohamov replied:
@ydamigos Unfortunate no. 
1) Handling of zero length breaks F4 procedure
2) Reading of 1 bytes lead to two actual bytes "read". At least clock is generated.
3) Reading of two bytes generates NACK after first byte, so slave send nothing. 

Code that is working properly seems to me a bit odd...
```
if (data->current.is_write) {
	LL_I2C_ClearFlag_ADDR(i2c);
	return;
}
if (data->current.len == 0U && IS_ENABLED(CONFIG_SOC_SERIES_STM32F1X)) {
	LL_I2C_GenerateStopCondition(i2c);
} else if (data->current.len == 1U) {
	/* Single byte reception: enable NACK and clear POS */
	LL_I2C_AcknowledgeNextData(i2c, LL_I2C_NACK);
#ifdef CONFIG_SOC_SERIES_STM32F1X
	LL_I2C_ClearFlag_ADDR(i2c);
	LL_I2C_GenerateStopCondition(i2c);
#endif
} else if (data->current.len == 2U) {
#ifdef CONFIG_SOC_SERIES_STM32F1X
	LL_I2C_ClearFlag_ADDR(i2c);
#endif
	/* 2-byte reception: enable NACK and set POS */
	LL_I2C_AcknowledgeNextData(i2c, LL_I2C_NACK);
	LL_I2C_EnableBitPOS(i2c);
}
LL_I2C_ClearFlag_ADDR(i2c);
```

pavlohamov commented:
@avisconti I've updated the description. Could you be so kind to take a look?

pavlohamov commented:
@avisconti I've added reference to the Errata section. Also i've updated the handling code according to @ydamigos comments. Could you please take a look?

avisconti commented:
> @avisconti I've added reference to the Errata section. Also i've updated the handling code according to @ydamigos comments. Could you please take a look?

Sure, but I'll do it next week.

ydamigos commented:
@pavlohamov STM32F103 reference manual [RM0008](https://www.st.com/content/ccc/resource/technical/document/reference_manual/59/b9/ba/7f/11/af/43/d5/CD00171190.pdf/files/CD00171190.pdf/jcr:content/translations/en.CD00171190.pdf) describes two methods for closing communication in master receiver mode depending on I2C interrupt priority. Could you test if we decrease the interrupt priority (it is 0 right now) if I2C will work with the same code for F1 and F4 series?

```
Method 1: This method is for the case when the I2C is used with interrupts that have the
highest priority in the application.
The master sends a NACK for the last byte received from the slave. After receiving this
NACK, the slave releases the control of the SCL and SDA lines. Then the master can send
a Stop/Restart condition.

1. To generate the nonacknowledge pulse after the last received data byte, the ACK bit must be cleared just after reading the second last data byte (after second last RxNE event).
2. To generate the Stop/Restart condition, software must set the STOP/START bit just after reading the second last data byte (after the second last RxNE event).
3. In case a single byte has to be received, the Acknowledge disable and the Stop condition generation are made just after EV6 (in EV6_1, just after ADDR is cleared).

Method 2: This method is for the case when the I2C is used with interrupts that do not have the highest priority in the application or when the I2C is used with polling.With this method, DataN_2 is not read, so that after DataN_1, the communication is stretched (both RxNE and BTF are set). Then, clear the ACK bit before reading DataN-2 in DR to ensure it is be cleared before the DataN Acknowledge pulse. After that, just after reading DataN_2, set the STOP/ START bit and read DataN_1. After RxNE is set, read DataN.

When 3 bytes remain to be read:
•RxNE = 1 => Nothing (DataN-2 not read).
•DataN-1 received
•BTF = 1 because both shift and data registers are full: DataN-2 in DR and DataN-1 in the shift register => SCL tied low: no other data will be received on the bus.
•Clear ACK bit
•Read DataN-2 in DR => This will launch the DataN reception in the shift register
•DataN received (with a NACK)
•Program START/STOP
•Read DataN-1
•RxNE = 1
•Read DataN

The cases where a single byte or two bytes are to be received should be handled differently, as described below:
Case of a single byte to be received:
- In the ADDR event, clear the ACK bit.
- Clear ADDR
- Program the STOP/START bit
- Read the data after the RxNE flag is set.

Case of two bytes to be received:
- Set POS and ACK
- Wait for the ADDR flag to be set
- Clear ADDR
- Clear ACK
- Wait for BTF to be set
- Program STOP
- Read DR twice
```

pavlohamov commented:
@ydamigos Unfortunate right now i'm on business trip and have no HW with me. I'll return on 26 Aug.
Once i return, i will try to experiment with interrupt prio.

avisconti commented:
@pavlohamov 
The bug #17831 (which this PR is fixing) is still open. 
I think it is necessary to work on this PR in order to merge it in the zephyr 2.0 release.

@galak 
Do we still have time to include this fix in the upcoming version?

ioannisg commented:
@avisconti this needs to be moved to 2.1, since the review was not completed in time for 2.0 release.

avisconti commented:
> @avisconti this needs to be moved to 2.1, since the review was not completed in time for 2.0 release.

OK, but the correspondent fix should be issued in 2.0 later on. Is v2.0 a LTS?

erwango commented:
@ioannisg, is that possible to backport stuff on 2.0?

ioannisg commented:
> 
> 
> @ioannisg, is that possible to backport stuff on 2.0?

@avisconti @erwango there will be a 2.0.1 milestone, for a 2.0 maintenance release.

pavlohamov commented:
Hi all. Sorry for delayed answer. Cant get my head out of ongoing activities. 
@avisconti right now this fix could be assumed as finished. I want to solve this by updating the whole Event handing procedure by reading the SR register once and then using it instead of reading the SR each time. 
I've made some experiments: changing the IRQ priority, disabling interrupts in the handler (with lower prio).... And the conclusion is: this solution is most stable (not being affected by optimization changes).

Unfortunate cant really say when i will have time to dig deeper

avisconti commented:
> Hi all. Sorry for delayed answer. Cant get my head out of ongoing activities.
> @avisconti right now this fix could be assumed as finished. I want to solve this by updating the whole Event handing procedure by reading the SR register once and then using it instead of reading the SR each time.
> I've made some experiments: changing the IRQ priority, disabling interrupts in the handler (with lower prio).... And the conclusion is: this solution is most stable (not being affected by optimization changes).
> 
> Unfortunate cant really say when i will have time to dig deeper

Thanks very much.
I think that this commit should go in both v2.1.0 and v2.0.1

ydamigos commented on drivers/i2c/i2c_ll_stm32_v1.c:
```diff
@@ -210,24 +210,32 @@ static inline void handle_addr(struct device *dev)
+	 * specific way.
+	 * Please ref to STM32F10xxC/D/E I2C peripheral Errata sheet 2.14.1
+	 */
+	if (data->current.len == 0U && IS_ENABLED(CONFIG_SOC_SERIES_STM32F1X)) {
```
Do we need to send a STOP only for STM32F1X family when `len == 0`?

pavlohamov replied:
In previous implementation (before my changes) stop wasn't generated at all here. Thus i suppose it not required for other families. But for the F1, of stop is not generated here - no other vent will occur on 0-size transition

ioannisg commented on drivers/i2c/i2c_ll_stm32_v1.c:
```diff
@@ -210,24 +210,32 @@ static inline void handle_addr(struct device *dev)
+	 * specific way.
+	 * Please ref to STM32F10xxC/D/E I2C peripheral Errata sheet 2.14.1
+	 */
+	if (data->current.len == 0U && IS_ENABLED(CONFIG_SOC_SERIES_STM32F1X)) {
```
maybe add parenthesis in the first expression in the if statement

pavlohamov replied:
@ioannisg Sorry, missed the question. I would say it just a corner case for the F1 platform, because nothing no interrupt follows 0 length transaction

ioannisg commented:
> 
> 
> @ioannisg, is that possible to backport stuff on 2.0?

@erwango for this to be back-ported to a possible 2.0.1 release, the corresponding label should be added.

avisconti commented:
@erwango 
Are you directly taking care of the backporting to v2.0.1?
